### PR TITLE
Deprecated undocumented and unused `>` clear modifier.

### PR DESCRIPTION
**Pull request by Zimmi48**
The keep/clear modifier `>` is undocumented and **apparently unused in the whole CI** (tested by Jim in #15015). This feature does not look important enough to me to deserve a one-character shorthand, so instead of promoting it by documenting it, I propose to deprecate it in Coq 8.17 and remove it in Coq 8.18.

Note that while I can see the point of the clear modifier, with `apply` or `rewrite` (but it is easily replaced by appropriate calls to `clear`), I can't even find a relevant use case in the "keep" case (e.g., with `destruct`). If there is one, please let me know.

cc @herbelin who introduced the feature in 262e3151ce483aaab3b4f60e3d1dbdf875ea05ae (2014). BTW, at that time, the keep modifier was `!`, which according to 387351b4c0ffeff65d8a7192f5073cfd4bd20f53 was replaced by the use of parentheses around the argument. So I wonder why `>` ends up being supported in the `destruct` / `induction` case.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.

### CODE REVIEW

**Comment by herbelin** [comment]
> Note that while I can see the point of the clear modifier, with apply or rewrite (but it is easily replaced by appropriate calls to clear), I can't even find a relevant use case in the "keep" case (e.g., with destruct). If there is one, please let me know.

I would say that, normally, there should be no need to keep when `destruct` decide that it can be erased. However, `destruct` is sometimes too eager to erase.

Here is an example where, because of a non-variable index, `H` carries non redundant information to keep, but `destruct` nevertheless erases `H` by default:
```
Goal forall n (H:n = 0), H = H.
destruct H. 
```

> So I wonder why > ends up being supported in the destruct / induction case.

A situation like this one looks legitimate:
```
Goal forall f, f 0 = 0.
destruct >f.
```
(the effect of `>` is however not yet implemented)

**Comment by Zimmi48** [comment]
In your first example, is there a difference between what `destruct >H` and `destruct (H)` do?
In the second example, I fail to see what you have in mind that `destruct >H` should do.

**Comment by herbelin** [comment]
In the second example, `destruct >f` would force clearing `f`.

In the first example, `destruct (H)` keeps `H` and `destruct >H` should clear `H` (so doing the same as `destruct H`).

However, the code for `destruct >x` is apparently not activated.

PS: I'm not sure `>` is an intuitive enough notation. The idea is nevertheless that the linear use is part of the tactic (in contrast with using `clear` which would come as a separate tactic). Morally, in full words, one would expect something like `apply linear H`, or `apply consuming H`.

**Comment by JasonGross** [comment]
I think `destruct >H` might be slightly nicer than `destruct (H); clear H` as a way of ensuring that `destruct` clears section variables.

**Comment by Zimmi48** [comment]
I'm confused. I thought that in the `destruct` case, `>` was a "keep" modifier, but from your messages it looks like it is a "clear" modifier in all cases?

**Comment by herbelin** [comment]
> I'm confused. I thought that in the `destruct` case, `>` was a "keep" modifier, but from your messages it looks like it is a "clear" modifier in all cases?

To keep, the syntax was initially `!H` and moved to `(H)`.

The history is a bit confusing. The '>' was added to all cases of `rewrite` and to `destruct` in 1aecaf88e549 (15 Dec 2015) but without being implemented for `destruct`.

**Comment by silene** [comment]
What a waste. Clearing hypotheses might be the feature I miss the most from SSReflect when I am using Vanilla Coq, and now I learn it has actually existed for years.

For reference, the syntax in SSReflect is `rewrite {}H`. But note that it is more expressive than `rewrite >H`, since one can also write `rewrite {x}H`, assuming rewriting with `H` has removed any remaining occurrence of `x`.

**Comment by Zimmi48** [comment]
So what should we do with this feature? It looks like it is actually useful, so should we document it as is? Should we refine it to make it closer to SSReflect? I'm also of the opinion that the error message which talks about "keep/clear modifier" should be transformed to only talk about the "clear modifier" if the "keep modifier" is actually something different.

In any case, I think that for 8.17, the feature should either be documented (as is or refined) or removed (if we do not want to document it as is and do not have time to refine it). The "no undocumented feature" rule should continue to be enforced.

**Comment by Alizter** [comment]
I'm for refining it to be more like ssreflect.

**Comment by silene** [comment]
> It looks like it is actually useful, so should we document it as is?

I don't think so. It would have made sense to do so in 2014. But now that SSReflect is part of Coq, documenting a conflicting syntax does not seem like a good idea. Better remove it in my opinion. (To clarify my earlier comment about it being a waste, I was referring to the fact that this feature had never been advertised, not to the fact that it was being proposed for deprecation.)

File: plugins/ltac/g_tactic.mlg
Comment (SkySkimmer) [inline plugins/ltac/g_tactic.mlg:235]:
```suggestion
      { deprecated_keep_clear_modifier (); (* 8.17 *)
        (Some true,c) }
```

File: plugins/ltac/g_tactic.mlg
Comment (SkySkimmer) [inline plugins/ltac/g_tactic.mlg:194]:
```suggestion
    (fun () -> Pp.strbrk "The undocumented keep/clear modifier \">\" is deprecated. Open an issue at https://github.com/coq/coq/issues/new if you actually depend on this feature in your work.")
```

**Review by SkySkimmer — APPROVED** [review]
Will merge soon after my comments are answered if no opposition

**Comment by Zimmi48** [comment]
Will update to include a changelog.

**Comment by Zimmi48** [comment]
Done.

**Review by Alizter — APPROVED** [review]

**Comment by jfehrle** [comment]
I only learned of this PR this morning.

> > It looks like it is actually useful, so should we document it as is?

> I don't think so. It would have made sense to do so in 2014. But now that SSReflect is part of Coq, documenting a conflicting syntax does not seem like a good idea. Better remove it in my opinion. (To clarify my earlier comment about it being a waste, I was referring to the fact that this feature had never been advertised, not to the fact that it was being proposed for deprecation.)

That's not entirely persuasive.  From my limited perspective, I'm not aware of any effort or plans to make SSReflect and original tactics more consistent.  Furthermore, the documented syntax for SSReflect is _very_ out of date, and @gares told me earlier this year that no SSReflect expert he spoke to was willing to spend the time to help me make that chapter accurate--and it appeared it would be quite a lot of work both for doc writer(s) and reviewers to do a good job.

Absent a strategy here, what's the point of dropping the feature?  But if you guys want to deprecate it, I'm not going to stand in the way--though some response to this comment would be appreciate.

**Comment by jfehrle** [comment]
BTW, who is the release manager for 8.17?  Perhaps @SkySkimmer?

**Comment by gares** [comment]
While I supported the tool to check that the parser and the doc are in sync, I'm totally against using it to daclare what is properly documented and what is not. The fact nobody volunteer to rework so much doc is that they see too little added value to justify the work.

I'm in favor of dropping a feature which is not documented. FTR >f is valid ssr syntax, and it does not do a clear, so documenting it would just make more confusion arise.

**Comment by jfehrle** [comment]
> While I supported the tool to check that the parser and the doc are in sync, I'm totally against using it to daclare what is properly documented and what is not.

I didn't suggest that or even mention the tool.  It's just that it appeared to me that the documentation doesn't match the implemented grammar, though TBH I couldn't tell for sure.

**Comment by Zimmi48** [comment]
@jfehrle The SSR syntax uses many hacks, so I don't think you can judge based on your impression of how good the documentation reflects the underlying code. There may be gaps, but I don't think it can be qualified as "very out of date". Most likely the differences between the implemented and the documented syntax have always been there (neither has changed much in recent years).

Regarding this particular feature, this conversation has highlighted that the feature is unfinished. I proposed two options: refine and document or drop, with a deadline at the next release. Since it appears that there will not be any action to refine the feature by this time, it should be removed. Note that the situation would be *very* different if the feature already had users. But if it hasn't, we should regard it the same as a PR that we wouldn't merged because it is still in progress.

File: doc/changelog/04-tactics/16407-deprecate-keep-clear.rst
Comment (Zimmi48) [reply doc/changelog/04-tactics/16407-deprecate-keep-clear.rst:4]: I think I will keep things worded the way they are. The goal is just to give one chance to users to speak up if they exist.

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**
